nemo/collections/asr/parts/k2/maploss.py

```diff
@@ -0,0 +1,269 @@
+        lm_graph = self.lm_graph.clone()
+        if hasattr(lm_graph, "aux_labels"):
+            delattr(lm_graph, "aux_labels")
+        if self.pad_fsavec:
+            shift_labels_inpl([lm_graph], 1)
+        if self.loss_type == "mmi":
+            from nemo.collections.asr.parts.k2.graph_compilers import MmiTrainingGraphCompiler as compiler
+        else:
+            raise ValueError(f"Invalid value of `loss_type`: {self.loss_type}.")
+        self.graph_compiler = compiler(self.num_classes, self.topo_type, self.topo_with_selfloops, aux_graph=lm_graph,)
+
+    def _intersect_calc_scores_mmi_exact(
+        self, dense_fsa_vec: k2.DenseFsaVec, num_graphs: k2.Fsa, den_graph: k2.Fsa, return_lats: bool = True,
+    ):
+        device = dense_fsa_vec.device
+        assert device == num_graphs.device and device == den_graph.device
+
+        num_fsas = num_graphs.shape[0]
+        assert dense_fsa_vec.dim0() == num_fsas
+
+        den_graph = den_graph.clone()
+        num_graphs = num_graphs.clone()
+
+        num_den_graphs = k2.cat([num_graphs, den_graph])
+
+        # NOTE: The a_to_b_map in k2.intersect_dense must be sorted
+        # so the following reorders num_den_graphs.
+
+        # [0, 1, 2, ... ]
+        num_graphs_indexes = torch.arange(num_fsas, dtype=torch.int32)
+
+        # [num_fsas, num_fsas, num_fsas, ... ]
+        den_graph_indexes = torch.tensor([num_fsas] * num_fsas, dtype=torch.int32)
+
+        # [0, num_fsas, 1, num_fsas, 2, num_fsas, ... ]
+        num_den_graphs_indexes = torch.stack([num_graphs_indexes, den_graph_indexes]).t().reshape(-1).to(device)
+
+        num_den_reordered_graphs = k2.index_fsa(num_den_graphs, num_den_graphs_indexes)
+
+        # [[0, 1, 2, ...]]
+        a_to_b_map = torch.arange(num_fsas, dtype=torch.int32).reshape(1, -1)
+
+        # [[0, 1, 2, ...]] -> [0, 0, 1, 1, 2, 2, ... ]
+        a_to_b_map = a_to_b_map.repeat(2, 1).t().reshape(-1).to(device)
+
+        num_den_lats = k2.intersect_dense(
+            a_fsas=num_den_reordered_graphs,
+            b_fsas=dense_fsa_vec,
+            output_beam=10.0,
+            a_to_b_map=a_to_b_map,
+            seqframe_idx_name="seqframe_idx" if return_lats else None,
+        )
+
+        num_den_tot_scores = num_den_lats.get_tot_scores(log_semiring=True, use_double_scores=True)
+        num_tot_scores = num_den_tot_scores[::2]
+        den_tot_scores = num_den_tot_scores[1::2]
+
+        if return_lats:
+            lat_slice = torch.arange(num_fsas, dtype=torch.int32).to(device) * 2
+            return (
+                num_tot_scores,
+                den_tot_scores,
+                k2.index_fsa(num_den_lats, lat_slice),
+                k2.index_fsa(num_den_lats, lat_slice + 1),
+            )
+        else:
+            return num_tot_scores, den_tot_scores, None, None
+
+    def _intersect_calc_scores_mmi_pruned(
+        self, dense_fsa_vec: k2.DenseFsaVec, num_graphs: k2.Fsa, den_graph: k2.Fsa, return_lats: bool = True,
```

Comment:
Same as above.

Reply:
Done in [7b2b4a4](https://github.com/NVIDIA/NeMo/pull/3351/commits/7b2b4a4ba4e34848d526d8c4ac53903bfbfe2fd2)